dd-sdk-android-ndk/src/main/cpp/utils/signal-monitor.c

```diff
@@ -152,61 +146,127 @@ bool configure_signal_stack() {
+    // block it and start a new thread for signals handling. When starting a new thread
+    // from main thread the new one will take the signals mask of the parent.
+    sigaddset(&signals_mask, SIGQUIT);
+    if (pthread_sigmask(SIG_BLOCK, &signals_mask, NULL) == 0) {
+
+        if (pthread_create(&datadog_watchdog_thread, NULL,
+                           override_native_signal_handlers, NULL) != 0) {
+            success = false;
+            __android_log_write(ANDROID_LOG_ERROR,
+                                LOG_TAG,
+                                "Was not able to create the watchdog thread");
+        }
 
-                previous_handler(signum);
-            }
+        // we restore the defaults on the main thread
+        if (pthread_sigmask(SIG_UNBLOCK, &signals_mask, NULL) != 0) {
+            success = false;
+            __android_log_write(ANDROID_LOG_ERROR,
+                                LOG_TAG,
+                                "Was not able to restore the mask on SIGQUIT signal");
         }
+
+    } else {
+        success = false;
+        __android_log_write(ANDROID_LOG_ERROR,
+                            LOG_TAG,
+                            "Was not able to mask SIGQUIT signal");
     }
 
-    pthread_mutex_unlock(&handlerMutex);
+    return success;
 }
 
-void handle_signal(int signum, siginfo_t *info,
-                   void *user_context) {
-    int is_initialized = 0;
-    pthread_mutex_lock(&handlerMutex);
-    is_initialized = (global_env == NULL);
-    pthread_mutex_unlock(&handlerMutex);
-    if (is_initialized) return;
+bool try_to_install_handlers() {
+    if (handlers_installed) {
+        return true;
+    }
+
+    return (configure_signal_stack() &&
+            configure_global_sigaction() &&
+            start_watchdog_thread());
+}
+
+
+bool install_signal_handlers() {
+    pthread_mutex_lock(&mutex);
+    bool installed = try_to_install_handlers();
+    pthread_mutex_unlock(&mutex);
+    return installed;
+}
+
+void uninstall_signal_handlers() {
+    pthread_mutex_lock(&mutex);
+
+    if (!handlers_installed) {
+        pthread_mutex_unlock(&mutex);
+        return;
+    }
 
     const size_t signals_array_size = handled_signals_size();
     for (int i = 0; i < signals_array_size; i++) {
-        const int signal = handled_signals[i].signal_value;
-        if (signal == signum) {
-            crash_signal_intercepted(signal,
-                                     handled_signals[i].signal_name,
-                                     handled_signals[i].signal_error_message);
-            break;
+        struct sigaction *prev_action = &global_previous_sigaction[i];
+        if (prev_action) {
+            const int signal = handled_signals[i].signal_value;
+            sigaction(signal, prev_action, 0);
         }
     }
-    uninstall_signal_handlers();
-    invoke_previous_handler(signum, info, user_context);
+    if (global_sigaction != NULL) {
+        free(global_sigaction);
```

Comment:
I would set `global_sigaction` to `NULL` after the `free()` to make sure nobody uses it after it was released (and same below for `global_previous_sigaction`).